File: src/main/java/gargoyle/netsquares/prop/simple/BeanProperty.java
package gargoyle.netsquares.prop.simple;

import gargoyle.netsquares.prop.base.BaseRWProperty;

import java.beans.BeanInfo;
import java.beans.IntrospectionException;
import java.beans.Introspector;
import java.beans.PropertyDescriptor;
import java.util.Objects;

public final class BeanProperty<B, T> extends BaseRWProperty<T> {
    private final B bean;
    private final PropertyDescriptor propertyDescriptor;
    private final String propertyName;

    public BeanProperty(B bean, String propertyName) {
        this.bean = Objects.requireNonNull(bean);
        this.propertyName = Objects.requireNonNull(propertyName);
        BeanInfo beanInfo;
        try {
            beanInfo = Introspector.getBeanInfo(bean.getClass());
        } catch (IntrospectionException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
        for (PropertyDescriptor propertyDescriptor : beanInfo.getPropertyDescriptors()) {
            if (Objects.equals(propertyName, propertyDescriptor.getName())) {
                this.propertyDescriptor = propertyDescriptor;
                return;
            }
        }
        throw new IllegalArgumentException(propertyName);
    }

    @SuppressWarnings("unchecked")
    @Override
    public T get() {
        try {
            return (T) propertyDescriptor.getReadMethod().invoke(bean);
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    public B getBean() {
        return bean;
    }

    public String getPropertyName() {
        return propertyName;
    }

    @Override
    protected void _set(T value) {
        try {
            propertyDescriptor.getWriteMethod().invoke(bean, value);
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/simple/BeanProperty.kt
package gargoyle.netsquares.prop.simple

import gargoyle.netsquares.prop.base.BaseRWProperty
import java.beans.BeanInfo
import java.beans.Introspector
import java.beans.PropertyDescriptor

class BeanProperty<B : Any, T : Any>(val bean: B, val propertyName: String) : BaseRWProperty<T>() {
    private lateinit var propertyDescriptor: PropertyDescriptor

    init {
        val beanInfo: BeanInfo = Introspector.getBeanInfo(bean.javaClass)
        var ok = false
        for (propertyDescriptor in beanInfo.propertyDescriptors) {
            if (propertyName == propertyDescriptor.name) {
                this.propertyDescriptor = propertyDescriptor
                ok = true
            }
        }
        if (!ok) {
            error(propertyName)
        }
    }

    override fun _set(value: T) {
        try {
            propertyDescriptor.writeMethod.invoke(bean, value)
        } catch (e: ReflectiveOperationException) {
            throw RuntimeException(e.message, e)
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun get(): T {
        return try {
            propertyDescriptor.readMethod.invoke(bean) as T
        } catch (e: ReflectiveOperationException) {
            throw RuntimeException(e.message, e)
        }
    }
}
